contracts/modules/governance/NXMaster.sol

```diff
@@ -47,152 +50,173 @@ contract NXMaster is INXMMaster, Governed {
+    bytes2 contractCode,
+    address payable contractAddress,
     uint _type
-  )
-  external
-  onlyAuthorizedToGovern {
-    require(allContractVersions[_contractName] == address(0), "Contract code is already available.");
-    require(_contractAddress != address(0), "NULL address is not allowed.");
-    allContractNames.push(_contractName);
-    address newInternalContract = _contractAddress;
-    if (_type == 1) {
-      isUpgradable[_contractName] = true;
-    } else if (_type == 2) {
-      newInternalContract = _generateProxy(_contractAddress);
-      isProxy[_contractName] = true;
-    }
-    allContractVersions[_contractName] = address(uint160(newInternalContract));
-    contractsActive[newInternalContract] = true;
-    LegacyMasterAware up = LegacyMasterAware(allContractVersions[_contractName]);
-    up.changeMasterAddress(address(this));
-    up.changeDependentContractAddress();
-  }
+  ) internal {
 
-  /**
-   * @dev Anyone can close a claim if oraclize fails to close it.
-   * @param _claimId id of claim to be closed.
-   */
-  function closeClaim(uint _claimId) external {
+    require(contractAddresses[contractCode] == address(0), "NXMaster: code already in use");
+    require(contractAddress != address(0), "NXMaster: contract address is 0");
 
-    require(canCall(_claimId), "Payout retry time not reached.");
-    IClaimsReward cr = IClaimsReward(getLatestAddress("CR"));
-    cr.changeClaimStatus(_claimId);
-  }
+    contractCodes.push(contractCode);
 
-  function getOwnerParameters(bytes8 code) external view returns (bytes8 codeVal, address val) {
-    codeVal = code;
-    IQuotationData qd;
-    LegacyPoolData pd;
-    if (code == "MSWALLET") {
-      ITokenData td;
-      td = ITokenData(getLatestAddress("TD"));
-      val = td.walletAddress();
+    address newInternalContract;
+    if (_type == uint(ContractType.Replaceable)) {
 
-    } else if (code == "MCRNOTA") {
+      newInternalContract = contractAddress;
+      isReplaceable[contractCode] = true;
+    } else if (_type == uint(ContractType.Proxy)) {
 
-      pd = LegacyPoolData(getLatestAddress("PD"));
-      val = pd.notariseMCR();
+      newInternalContract = address(new OwnedUpgradeabilityProxy(contractAddress));
+      isProxy[contractCode] = true;
+    } else {
+      revert("NXMaster: Unsupported contract type");
+    }
 
-    } else if (code == "OWNER") {
+    contractAddresses[contractCode] = address(uint160(newInternalContract));
+    contractsActive[newInternalContract] = true;
 
-      val = owner;
+    MasterAware up = MasterAware(newInternalContract);
+    up.changeMasterAddress(address(this));
+    up.changeDependentContractAddress();
 
-    } else if (code == "QUOAUTH") {
+    emit InternalContractAdded(contractCode, contractAddress, ContractType(_type));
+  }
 
-      qd = IQuotationData(getLatestAddress("QD"));
-      val = qd.authQuoteEngine();
+  /// @dev upgrades multiple contracts at a time
+  function upgradeMultipleContracts(
```

Comment:
Same as above, the function doesn't match INXMaster and it could be external and use calldata instead.

Reply:
Given that it was public the function selector might need to be updated to the new one in proposal categories.

Reply:
INXMaster needs to be like that cause it's an interface i think i'll leave this as is